api/v1alpha1/clusterconfig_types.go

```diff
@@ -344,6 +348,66 @@ func (ImageRegistries) VariableSchema() clusterv1.VariableSchema {
 	}
 }
 
+type Users []User
+
+func (Users) VariableSchema() clusterv1.VariableSchema {
+	return clusterv1.VariableSchema{
+		OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+			Description: "Users to add to the machine",
+			Type:        "array",
+			Items:       ptr.To(User{}.VariableSchema().OpenAPIV3Schema),
+		},
+	}
+}
+
+// User defines the input for a generated user in cloud-init.
+type User struct {
+	// Name specifies the user name
+	Name string `json:"name"`
+
+	// Passwd specifies a hashed password for the user
+	// +optional
+	Passwd *string `json:"passwd,omitempty"`
```

Comment:
for some optional fields we use just a string, and for this we're using a pointer. 

i forgot the reasoning behind not using a pointer, but this seems to be a convention in most projects. i'd like to get some consensus and land on what we're doing.

Reply:
I think this a good explanation:

> Using a pointer allows distinguishing unset from the zero value for that type. There are some cases where, in principle, a pointer is not needed for an optional field since the zero value is forbidden, and thus implies unset. There are examples of this in the codebase. However:
> 
> it can be difficult for implementors to anticipate all cases where an empty value might need to be distinguished from a zero value
> structs are not omitted from encoder output even where omitempty is specified, which is messy;
> having a pointer consistently imply optional is clearer for users of the Go language client, and any other clients that use corresponding types
> 
> Therefore, we ask that pointers always be used with optional fields that do not have a built-in nil value.
> -- https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required explains the need for using a pointer

Reply:
I see Passwd can be `""` which is not the same as having no password.

Reply:
That's what I thought at first... but it turns out that an empty string is invalid here, because it's eventually passed to cloud-init's `passwd` field, which requires the _hash_ of a password. And the hash can never be an empty string.

If we make this a string, rather than a pointer to a string, then the zero value (empty string) won't be marshalled, which is actually what we want.

Reply:
I'm not clear from the discussion above whether we want to keep this as a pointer or not? I don't think we need this to be a pointer as it has a valid zero value that will not be marshalled.

Reply:
> I don't think we need this to be a pointer as it has a valid zero value that will not be marshalled.

What I tried to say abovei s that the zero value is _not_ valid: An empty string cannot be a hashed password.

I'll change it from a pointer to a value.

CABPK supports both cloud-init and ignition for bootstrap, and both take a hashed password as input:
- https://github.com/kubernetes-sigs/cluster-api/blob/0624ec0790d1190fcb980c4816cd38aa430e4b87/bootstrap/kubeadm/internal/cloudinit/users.go#L24-L26
- https://github.com/kubernetes-sigs/cluster-api/blob/7bb4cde098319db9378a4f327294c96d645e892f/bootstrap/kubeadm/internal/ignition/clc/clc.go#L76-L78

Reply:
Changed to value.

Reply:
Perfect thank you!